core/src/banking_stage/transaction_scheduler/transaction_state.rs

```diff
@@ -0,0 +1,307 @@
+use {
+    solana_runtime::transaction_priority_details::TransactionPriorityDetails,
+    solana_sdk::{slot_history::Slot, transaction::SanitizedTransaction},
+};
+
+/// Simple wrapper type to tie a sanitized transaction to max age slot.
+pub(crate) struct SanitizedTransactionTTL {
+    pub(crate) transaction: SanitizedTransaction,
+    pub(crate) max_age_slot: Slot,
+}
+
+/// TransactionState is used to track the state of a transaction in the transaction scheduler
+/// and banking stage as a whole.
+///
+/// There are two states a transaction can be in:
+///     1. `Unprocessed` - The transaction is available for scheduling.
+///     2. `Pending` - The transaction is currently scheduled or being processed.
+///
+/// Newly received transactions are initially in the `Unprocessed` state.
+/// When a transaction is scheduled, it is transitioned to the `Pending` state,
+///   using the `transition_to_pending` method.
+/// When a transaction finishes processing it may be retryable. If it is retryable,
+///   the transaction is transitioned back to the `Unprocessed` state using the
+///   `transition_to_unprocessed` method. If it is not retryable, the state should
+///   be dropped.
+///
+/// For performance, when a transaction is transitioned to the `Pending` state, the
+///   internal `SanitizedTransaction` is moved out of the `TransactionState` and sent
+///   to the appropriate thread for processing. This is done to avoid cloning the
+///  `SanitizedTransaction`.
+#[allow(clippy::large_enum_variant)]
+pub(crate) enum TransactionState {
+    /// The transaction is available for scheduling.
+    Unprocessed {
+        transaction_ttl: SanitizedTransactionTTL,
+        transaction_priority_details: TransactionPriorityDetails,
+        forwarded: bool,
```

Comment:
I actually thought about that, the issue I think comes from when we move to `Pending`.
An `Unprocessed` transaction may be forwarded or not, depending on when we received it.
When we move to `Pending`, we need to keep track of whether or not it was forwarded in order to not forward it again.

Might be missing the suggestion, but I didn't see a way to do that without `PendingUnforwarded` and `PendingForwarded`, at which point I decided to just use a bool.

Reply:
I didn't think through this, was imagining the `Forwarder` keeps tracking forwarded packets, so if a packet goes around scheduler path but still ended at `Forwarder`, it'd check it against kept records and decided to forward or drop. So that way the enum will be free from `forwarded` flag and logic. But that might not be worth it, assuming it is doable.